releases: add tidb 5.0.6 release notes (#8041)
pingcap/docs-cn: releases/release-5.0.6.md

---
title: TiDB 5.0.6 Release Notes
---

# TiDB 5.0.6 Release Notes

发版日期：2021 年 12 月 28 日

TiDB 版本：5.0.6

## 兼容性更改

+ Tools

    + TiCDC

        - cdc server 命令错误输出从标准输出改成标准错误. [#3133](https://github.com/pingcap/tiflow/issues/3133)

## 提升改进

+ TiDB

    (dup) - 修复乐观事务冲突可能导致事务相互阻塞的问题 [#11148](https://github.com/tikv/tikv/issues/11148)
    (dup) - 修复运行 MPP 查询时出现 `invalid cop task execution summaries length` 相关日志的问题 [#1791](https://github.com/pingcap/tics/issues/1791)
    (dup) - 当 coprocessor 遇到锁时，在调试日志中显示受影响的 SQL 语句帮助诊断问题 [#27718](https://github.com/pingcap/tidb/issues/27718)

> + TiKV
> 
>     - 将插入 SST 文件时的校验操作移动到了 import 线程池以提高 SST 文件插入速度. [#11239](https://github.com/tikv/tikv/issues/11239)
>     - 为 raft 日志回收模块增加了更多监控以定位问题. [#11374](https://github.com/tikv/tikv/issues/11374)
>     - 折叠了部分 Storage 相关的不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)
> 
> + PD
> 
>     (dup) - 优化调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)
>     - 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4497](https://github.com/tikv/pd/issues/4497)

Reviewer: ```suggestion
+ TiKV

    - 将插入 SST 文件时的校验操作移动到了 import 线程池，从而提高 SST 文件的插入速度 [#11239](https://github.com/tikv/tikv/issues/11239)
    - 在 Raft 日志垃圾回收模块中添加了更多监控，从而定位该模块中出现的问题 [#11374](https://github.com/tikv/tikv/issues/11374)
    - 折叠了部分与 Storage 相关的不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)

+ PD

    (dup) - 优化调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)
    - 通过允许空 Region 调度和修复配置的方式，使 `scatter-range-scheduler` 调度器运行得更加顺畅 [#4497](https://github.com/tikv/pd/issues/4497)
```

Reviewer 2: 更加顺畅 改为 更加均匀 或 效果更好？

+ Tools

    + TiCDC

        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)
        (dup) - 修复当扫描存量数据耗时过长时，可能由于 TiKV 进行 GC 而导致存量数据扫描失败的问题 [#2470](https://github.com/pingcap/tiflow/issues/2470)
        (dup) - 修复当发生 ErrGCTTLExceeded 错误时，changefeed 不快速失败的问题 [#3111](https://github.com/pingcap/tiflow/issues/3111)
        (dup) - 为 EtcdWorker 添加 tick 频率限制，防止 PD 的 etcd 写入次数过于频繁影响 PD 服务 [#3112](https://github.com/pingcap/tiflow/issues/3112)
        - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)
        (dup) - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
        - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3352](https://github.com/pingcap/tiflow/issues/3352)

>         - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)

Reviewer 3: ```suggestion
```

>         - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)

Reviewer 3: ```
        - Kafka sink 模块的 `MaxMessageBytes` 默认值设置为 10MB，并使用 Kafka broker 参数 `message.max.bytes` 和 Topic 参数 `max.message.bytes` 的较小值对 producer 进行初始化[#3081](https://github.com/pingcap/tiflow/issues/3081)
```

Author: > ```
>         - Kafka sink 模块的 `MaxMessageBytes` 默认值设置为 10MB，并使用 Kafka broker 参数 `message.max.bytes` 和 Topic 参数 `max.message.bytes` 的较小值对 producer 进行初始化[#3081](https://github.com/pingcap/tiflow/issues/3081)
>         - 删除 TiCDC 内部 owner 和 processor 重构后遗留的相关历史代码 [#2765](https://github.com/pingcap/tiflow/issues/2765)
> ```

The notes should point out the benefits to users, instead of describing what we have done to the codes. Please optimize the notes. Thanks.

Reviewer 4: Does this involve configuration value change? If yes, please add it to the **Compatibility changes** part. Thanks!
@shichun-0415 @maxshuang

Reviewer 3: I have moved the configuration value change to Compatibility changes part.

        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#4054](https://github.com/pingcap/tiflow/issues/4054), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606)

    + (Backup & Restore) BR

        - 支持 pd request 和 TiKV IO 超时错误重试 [#27787](https://github.com/pingcap/tidb/issues/27787)

## Bug 修复

+ TiDB

>     (dup) - 修复 `grant` 和 `revoke` 操作在授予和撤销全局权限时，报 `privilege check fail` 错误的问题 [#29675](https://github.com/pingcap/tidb/issues/29675)

Author: ```suggestion
    - 修复 DML 和 DDL 语句并发执行时可能发生 panic 的问题 [#30940](https://github.com/pingcap/tidb/issues/30940)
    (dup) - 修复 `grant` 和 `revoke` 操作在授予和撤销全局权限时，报 `privilege check fail` 错误的问题 [#29675](https://github.com/pingcap/tidb/issues/29675)
```

    (dup) - 修复 MPP 节点的可用性检测在某些边界场景中无法工作的问题 [#3118](https://github.com/pingcap/tics/issues/3118)
    (dup) - 修复分配 `MPP task ID` 时出现 `DATA RACE` 的问题 [#27952](https://github.com/pingcap/tidb/issues/27952)
    (dup) - 修复删除空的 `dual table` 后 MPP 查询出现 `index out of range` 报错的问题 [#28250](https://github.com/pingcap/tidb/issues/28250)
    - 修复并行插入无效日期类型值时，TiDB panic 的问题 [#25393](https://github.com/pingcap/tidb/issues/25393)
    - 修复在 MPP 模式下查询时，报 `can not found column in Schema column` 错误的问题 [#30980](https://github.com/pingcap/tidb/issues/30980)
    (dup) - 修复 TiDB 在 TiFlash 关闭时可能出现 panic 的问题 [#28096](https://github.com/pingcap/tidb/issues/28096)
    - 修复优化器在进行 join reorder 优化时，报 `index out of range` 错误的问题 [#24095](https://github.com/pingcap/tidb/issues/24095)
    (dup) - 修复当 `ENUM` 类型作为 `IF` 或 `CASE WHEN` 等控制函数的参数时，返回结果不正确的问题 [#23114](https://github.com/pingcap/tidb/issues/23114)
    - 修复 `CONCAT(IFNULL(TIME(3))` 计算结果出错的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
    - 修复当 `GREATEST` 和 `LEAST`  函数传入无符号整型值时，计算结果出错的问题 [#30101](https://github.com/pingcap/tidb/issues/30101)

>     - 修复当 SQL 中存在 json 类型列 与 char 类型列 join 时，SQL 出错的问题 [#29401](https://github.com/pingcap/tidb/issues/29401)

Reviewer 4: ```suggestion
    - 修复当 SQL 语句中存在 JSON 类型列与 `CHAR` 类型列连接时，SQL 出错的问题 [#29401](https://github.com/pingcap/tidb/issues/29401)
```

>     - 修复 update 语句未更新索引处理优化，可能导致数据索引不一致的问题 [#30410](https://github.com/pingcap/tidb/issues/30410)

Reviewer 4: ```suggestion
    - 修复错误使用 lazy existence 检查和 untouched key optimization 可能导致数据索引不一致的问题 [#30410](https://github.com/pingcap/tidb/issues/30410)
```

>     - 修复窗口函数在使用事务时，计算结果与不使用事务不一样的问题 [#29947](https://github.com/pingcap/tidb/issues/29947)

Reviewer 4: ```suggestion
    - 修复窗口函数在使用事务时，计算结果与不使用事务时的计算结果不一致的问题 [#29947](https://github.com/pingcap/tidb/issues/29947)
```

>     - 修复 SQL 形如`cast(integer as char) union string` 计算结果出错的问题 [#29513](https://github.com/pingcap/tidb/issues/29513)

Reviewer 4: ```suggestion
    - 修复 SQL 语句中包含 `cast(integer as char) union string` 时计算结果出错的问题 [#29513](https://github.com/pingcap/tidb/issues/29513)
```

    (dup) - 修复将 `Decimal` 转为 `String` 时长度信息错误的问题 [#29417](https://github.com/pingcap/tidb/issues/29417)

>     - 修复使用 NATURAL JOIN 时，报错 `Column 'col_name' in field list is ambiguous` 的问题 [#25041](https://github.com/pingcap/tidb/issues/25041)

Reviewer 4: ```suggestion
    - 修复 SQL 语句中带有 NATURAL JOIN 时，执行意外报错 `Column 'col_name' in field list is ambiguous` 的问题 [#25041](https://github.com/pingcap/tidb/issues/25041)
```

    (dup) - 修复由于 `tidb_enable_vectorized_expression` 设置的值不同（`on` 或 `off`）导致 `GREATEST` 函数返回结果不一致的问题 [#29434](https://github.com/pingcap/tidb/issues/29434)
    (dup) - 修复在某些情况下 Planner 可能缓存无效 `join` 计划的问题 [#28087](https://github.com/pingcap/tidb/issues/28087)
    - 修复 SQL 在 join 上计算聚合函数时，报错 `index out of range [1] with length 1` 的问题 [#1978](https://github.com/pingcap/tics/issues/1978)

> + TiKV
> 
>     - 修复某个 TiKV 节点停机导致 Resolve Timestamp 进度落后的问题 [#11351](https://github.com/tikv/tikv/issues/11351)
>     (dup) - 修复 Raft client 中 batch 消息过大的问题 [#9714](https://github.com/tikv/tikv/issues/9714)
>     (dup) - 修复在极端情况下同时进行 Region Merge、ConfChange 和 Snapshot 时，TiKV 会出现 Panic 的问题 [#11475](https://github.com/tikv/tikv/issues/11475)
>     - 修复逆序范围查询时没有正确读到内存锁的问题 [#11440](https://github.com/tikv/tikv/issues/11440)
>     (dup) - 修复 Decimal 除法计算的结果为 0 时符号为负的问题 [#29586](https://github.com/pingcap/tidb/issues/29586)
>     (dup) - 修复历史版本回收任务堆积导致的内存耗尽问题 [#11410](https://github.com/tikv/tikv/issues/11410)
>     (dup) - 修复 TiKV 监控项中实例级别 gRPC 的平均延迟时间不准确的问题 [#11299](https://github.com/tikv/tikv/issues/11299)
>     - 修复 label 泄漏导致 OOM 的问题. [#11195](https://github.com/tikv/tikv/issues/11195)
>     (dup) - 修复在缺失下游数据库时出现 TiCDC Panic 的问题 [#11123](https://github.com/tikv/tikv/issues/11123)
>     (dup) - 修复因 Congest 错误而导致的 CDC 频繁增加 scan 重试的问题 [#11082](https://github.com/tikv/tikv/issues/11082)
>     (dup) - 修复因 channel 打满而导致的 Raft 断连的问题 [#11047](https://github.com/tikv/tikv/issues/11047)
>     - 修复 Lightning 导入数据期间 TiKV 因为文件不存在而宕机的问题 [#10438](https://github.com/tikv/tikv/issues/10438)
>      (dup) - 修复由于无法在 `Max`/`Min` 函数中正确识别 Int64 是否为有符号整数，导致  `Max`/`Min` 函数的计算结果不正确的问题 [#10158](https://github.com/tikv/tikv/issues/10158)
>     - 修复从副本从主副本获取快照后未能正确修改元信息导致 TiKV 节点宕机的问题。  [#10225](https://github.com/tikv/tikv/issues/10225)
>     - 修复 backup 线程池泄漏的问题 [#10287](https://github.com/tikv/tikv/issues/10287)
>     - 修复将不合法字符串转化为浮点数的问题 [#23322](https://github.com/pingcap/tidb/issues/23322)

Reviewer: ```suggestion
+ TiKV

    - 修复某个 TiKV 节点停机会导致 Resolved Timestamp 进度落后的问题 [#11351](https://github.com/tikv/tikv/issues/11351)
    (dup) - 修复 Raft client 中 batch 消息过大的问题 [#9714](https://github.com/tikv/tikv/issues/9714)
    (dup) - 修复在极端情况下同时进行 Region Merge、ConfChange 和 Snapshot 时，TiKV 会出现 Panic 的问题 [#11475](https://github.com/tikv/tikv/issues/11475)
    - 修复逆序范围查询时没有正确读到内存锁的问题 [#11440](https://github.com/tikv/tikv/issues/11440)
    (dup) - 修复 Decimal 除法计算的结果为 0 时符号为负的问题 [#29586](https://github.com/pingcap/tidb/issues/29586)
    (dup) - 修复历史版本回收任务堆积导致的内存耗尽问题 [#11410](https://github.com/tikv/tikv/issues/11410)
    (dup) - 修复 TiKV 监控项中实例级别 gRPC 的平均延迟时间不准确的问题 [#11299](https://github.com/tikv/tikv/issues/11299)
    - 修复 label 泄漏导致内存溢出 (OOM) 的问题 [#11195](https://github.com/tikv/tikv/issues/11195)
    (dup) - 修复在缺失下游数据库时出现 TiCDC Panic 的问题 [#11123](https://github.com/tikv/tikv/issues/11123)
    (dup) - 修复因 Congest 错误而导致的 CDC 频繁增加 scan 重试的问题 [#11082](https://github.com/tikv/tikv/issues/11082)
    (dup) - 修复因 channel 打满而导致的 Raft 断连的问题 [#11047](https://github.com/tikv/tikv/issues/11047)
    - 修复在 Lightning 导入数据的期间 TiKV 会因为文件不存在而出现 panic 的问题 [#10438](https://github.com/tikv/tikv/issues/10438)
     (dup) - 修复由于无法在 `Max`/`Min` 函数中正确识别 Int64 是否为有符号整数，导致  `Max`/`Min` 函数的计算结果不正确的问题 [#10158](https://github.com/tikv/tikv/issues/10158)
    - 修复从主副本 (leader) 获取快照后因未能正确地修改元信息而导致 TiKV 节点 panic 的问题 [#10225](https://github.com/tikv/tikv/issues/10225)
    - 修复 backup 线程池泄漏的问题 [#10287](https://github.com/tikv/tikv/issues/10287)
    - 修复将不合法字符串转化为浮点数的问题 [#23322](https://github.com/pingcap/tidb/issues/23322)
```

Reviewer: 请问这里的第 97 行“修复 label 泄漏导致内存溢出 (OOM) 的问题 [#11195](https://github.com/tikv/tikv/issues/11195)”与 [5.3 release notes](https://docs.pingcap.com/zh/tidb/stable/release-5.3.0/#bug-%E4%BF%AE%E5%A4%8D) 中的“修复因统计线程监控数据导致的内存泄漏 #11195” 相同吗？

Reviewer 5: LGTM。
最后一条英文的 release note 以中文为准。 我不太确定 copr 是否需要在 release note中提及。real 在这里的意思就是“浮点数” 而不是修饰 bug fix。

> + PD
> 
>     (dup) - 修复 TiKV 节点缩容后可能导致 Panic 的问题 [#4344](https://github.com/tikv/pd/issues/4344)
>     (dup) - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)
>     (dup) - 修复因 Region syncer 卡住而导致 leader 选举慢的问题 [#3936](https://github.com/tikv/pd/issues/3936)
>     (dup) - 允许 Evict Leader 调度器调度拥有不健康副本的 Region [#4093](https://github.com/tikv/pd/issues/4093)
>     - 修复 Down Store 的 Remove Peer 速度受限问题[#4090](https://github.com/tikv/pd/issues/4090)
>     - 修复当 Region 心跳低于60秒导致热点 Cache 不能清空的问题 [#4390](https://github.com/tikv/pd/issues/4390)

Reviewer: ```suggestion
+ PD

    (dup) - 修复 TiKV 节点缩容后可能导致 Panic 的问题 [#4344](https://github.com/tikv/pd/issues/4344)
    (dup) - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)
    (dup) - 修复因 Region syncer 卡住而导致 leader 选举慢的问题 [#3936](https://github.com/tikv/pd/issues/3936)
    (dup) - 允许 Evict Leader 调度器调度拥有不健康副本的 Region [#4093](https://github.com/tikv/pd/issues/4093)
    - 修复当对宕机的节点进行修复时 Remove Peer 的速度受限的问题 [#4090](https://github.com/tikv/pd/issues/4090)
    - 修复 Region 心跳低于 60 秒会导致热点 Cache 不能清空的问题 [#4390](https://github.com/tikv/pd/issues/4390)
```

Reviewer 2: lgtm

        - 修复 Avro sink 模块不支持 json type 列类型解析 [#3624](https://github.com/pingcap/tiflow/issues/3624)
        - 修复 TiKV 重启时 cdc 读取到错误的元数据快照问题 [#2603](https://github.com/pingcap/tiflow/issues/2603)
        (dup) - 修复执行 DDL 后的内存泄漏的问题 [#3174](https://github.com/pingcap/tiflow/issues/3174)
        - 修复 Canal 和 Maxwell 协议没有自动开启 old value 选项问题 [#3676](https://github.com/pingcap/tiflow/issues/3676)
        - 修复 cdc server 在一些 RHEL 系统(6.8, 6.9 etc)上运行出现的时区错误问题.  [#3584](https://github.com/pingcap/tiflow/issues/3584)
        - 修复 Kafka sink 模块监控变量 txn_batch_size 不准确问题 [#3431](https://github.com/pingcap/tiflow/issues/3431)
         (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)
        (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)
        (dup) - 修复 tikv_cdc_min_resolved_ts_no_change_for_1m 监控在没有 changefeed 的情况下持续更新的问题 [#11017](https://github.com/tikv/tikv/issues/11017)
        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)

> 

Reviewer 3: ```suggestion

        - 修复 MQ Sink 在非 open-protocol 协议下初始化参数 `max-message-bytes`  导致的 `Message was too large` 错误
        - 修复人为删除 ETCD 任务状态时导致的 TiCDC panic 问题
```

---
title: TiDB 5.0.6 Release Notes
---

# TiDB 5.0.6 Release Notes

> 发版日期：2021 年 12 月 28 日

Author: ```suggestion
发版日期：2021 年 12 月 31 日
```

TiDB 版本：5.0.6

## 兼容性更改

+ Tools

    + TiCDC

>         - cdc server 命令错误输出从标准输出改成标准错误. [#3133](https://github.com/pingcap/tiflow/issues/3133)

Reviewer 6: ```suggestion
        - cdc server 命令错误输出从标准输出改为标准错误 [#3133](https://github.com/pingcap/tiflow/issues/3133)
```

## 提升改进

+ TiDB

    (dup) - 修复乐观事务冲突可能导致事务相互阻塞的问题 [#11148](https://github.com/tikv/tikv/issues/11148)
    (dup) - 修复运行 MPP 查询时出现 `invalid cop task execution summaries length` 相关日志的问题 [#1791](https://github.com/pingcap/tics/issues/1791)
    (dup) - 当 coprocessor 遇到锁时，在调试日志中显示受影响的 SQL 语句帮助诊断问题 [#27718](https://github.com/pingcap/tidb/issues/27718)

+ TiKV

    - 将插入 SST 文件时的校验操作从 Apply 线程池移动到 Import 线程池，从而提高 SST 文件的插入速度 [#11239](https://github.com/tikv/tikv/issues/11239)
    - 在 Raft 日志垃圾回收模块中添加了更多监控，从而定位该模块中出现的性能问题 [#11374](https://github.com/tikv/tikv/issues/11374)

>     - 折叠了 Grafana dashboard 中与 Storage 相关的部分不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)

Reviewer: ```suggestion
    - 折叠了 Grafana Dashboard 中与 Storage 相关的部分不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)
```

+ PD

    (dup) - 优化调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)
    - 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4497](https://github.com/tikv/pd/issues/4497)

+ Tools

    + TiCDC

        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)

>         - 减少 Sink 模块锁冲突 [#2760](https://github.com/pingcap/tiflow/pull/2760)

Reviewer 6: ```suggestion
        - 减少了 Sink 模块锁冲突 [#2760](https://github.com/pingcap/tiflow/pull/2760)
```

>         - 减少 Sink 模块锁冲突 [#2760](https://github.com/pingcap/tiflow/pull/2760)

Author: ```suggestion
```

        (dup) - 修复当发生 ErrGCTTLExceeded 错误时，changefeed 不快速失败的问题 [#3111](https://github.com/pingcap/tiflow/issues/3111)
        (dup) - 为 EtcdWorker 添加 tick 频率限制，防止 PD 的 etcd 写入次数过于频繁影响 PD 服务 [#3112](https://github.com/pingcap/tiflow/issues/3112)
        - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)
        (dup) - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
        - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3352](https://github.com/pingcap/tiflow/issues/3352)

>         - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)

Reviewer 3: ```suggestion
        - Kafka sink 模块的 `MaxMessageBytes` 默认值设置为 10MB，并优化 producer 参数`max.message.bytes` 初始值，降低用户设置 `message.max.bytes` 值过大导致的 Kafka 消息无法发送问题的概率 [#3081](https://github.com/pingcap/tiflow/issues/3081)
```

>         - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)
>         (dup) - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
>         - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3352](https://github.com/pingcap/tiflow/issues/3352)
>         - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)
>         - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#4054](https://github.com/pingcap/tiflow/issues/4054), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606)

Reviewer 6: ```suggestion
        - 支持消息 Batch 操作以减少 EtcdWorker tick [#3112](https://github.com/pingcap/tiflow/issues/3112)
        (dup) - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
        - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3352](https://github.com/pingcap/tiflow/issues/3352)
        - Kafka sink 模块的 `MaxMessageBytes` 默认值设置为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)
        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#4054](https://github.com/pingcap/tiflow/issues/4054), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606)
```

    + (Backup & Restore) BR

>         - 支持 pd request 和 TiKV IO 超时错误重试 [#27787](https://github.com/pingcap/tidb/issues/27787)

Reviewer 6: ```suggestion
        - 支持 PD request 和 TiKV IO 发生超时错误后进行重试 [#27787](https://github.com/pingcap/tidb/issues/27787)
```

Reviewer 2: LGTM

Reviewer: @hfxsd 这里的 IO 在 TiKV 模块写为 ”I/O"~

>         - 支持 PD request 和 TiKV I/O 发生超时错误后进行重试 [#27787](https://github.com/pingcap/tidb/issues/27787)

Author: ```suggestion
        - 遇到 PD 请求错误或 TiKV I/O 超时错误时重试 BR 任务 [#27787](https://github.com/pingcap/tidb/issues/27787)
```

> 

Author: ```suggestion

    + TiDB Lightning

        - 支持导入数据到带有表达式索引或带有基于虚拟生成列的索引的表中 [#1404](https://github.com/pingcap/br/issues/1404)
```

    (dup) - 修复因 channel 打满而导致的 Raft 断连的问题 [#11047](https://github.com/tikv/tikv/issues/11047)
    - 修复 Lightning 导入数据期间 TiKV 因为文件不存在而宕机的问题 [#10438](https://github.com/tikv/tikv/issues/10438)
     (dup) - 修复由于无法在 `Max`/`Min` 函数中正确识别 Int64 是否为有符号整数，导致  `Max`/`Min` 函数的计算结果不正确的问题 [#10158](https://github.com/tikv/tikv/issues/10158)
    - 修复从副本从主副本获取快照后未能正确修改元信息导致 TiKV 节点宕机的问题。  [#10225](https://github.com/tikv/tikv/issues/10225)
    - 修复 backup 线程池泄漏的问题 [#10287](https://github.com/tikv/tikv/issues/10287)
    - 修复将不合法字符串转化为浮点数的问题 [#23322](https://github.com/pingcap/tidb/issues/23322)

+ PD

    (dup) - 修复 TiKV 节点缩容后可能导致 Panic 的问题 [#4344](https://github.com/tikv/pd/issues/4344)
    (dup) - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)
    (dup) - 修复因 Region syncer 卡住而导致 leader 选举慢的问题 [#3936](https://github.com/tikv/pd/issues/3936)
    (dup) - 允许 Evict Leader 调度器调度拥有不健康副本的 Region [#4093](https://github.com/tikv/pd/issues/4093)
    - 修复 Down Store 的 Remove Peer 速度受限问题[#4090](https://github.com/tikv/pd/issues/4090)
    - 修复当 Region 心跳低于60秒导致热点 Cache 不能清空的问题 [#4390](https://github.com/tikv/pd/issues/4390)

+ TiFlash

>     - 修复当在扩大整数类型主键的列类型之后，可能会导致数据不一致的问题。
>     - 修复在某些平台 TiFlash 由于找不到 "libnsl.so" 而无法启动的问题，比如某些 ARM 平台
>     - 修复 TiFlash 的 Store size 统计信息与实际容量不一致的问题
>     - 修复 TiFlash 由于 "Cannot open file" 错误而导致的进程失败
>     - 修复当查询被取消时，TiFlash 偶发的崩溃问题。
>     - 修复查询报错 "3rd arguments of function substringUTF8 must be constants"
>     - 提高了支持的查询 SQL 中表达式的层级数量，避免比如过多 "OR" 条件导致查询报错
>     - 修复当查询带了 where <string> 会变成 int 类型的导致查询结构出错的问题。
>     - 修复在 TiFlash 与 TiDB/TiKV 之间，由于 CastStringAsDecimal 行为不一致的问题。
>     - 修复由于 "different types: expected Nullable(Int64), got Int64" 错误而导致查询报错的问题
>     - 修复由于 "Unexpected type of column: Nullable(Nothing)" 错误而导致的查询报错
>     - 修复 Decimal 类型比较时，可能导致数据溢出并导致查询失败的问题

Author: ```suggestion
    - 修复扩大整数类型主键的列类型之后，可能会导致数据不一致的问题
    - 修复在某些平台 TiFlash 由于找不到 "libnsl.so" 而无法启动的问题，比如某些 ARM 平台
    - 修复 TiFlash 的 Store size 统计信息与实际容量不一致的问题
    - 修复 TiFlash 由于 "Cannot open file" 错误而导致的进程失败
    - 修复当查询被取消时，TiFlash 偶发的崩溃问题
    - 修复查询报错 "3rd arguments of function substringUTF8 must be constants"
    - 修复由于过多 "OR" 条件导致的查询报错
    - 修复 where <string> 查询结果出错的问题
    - 修复在 TiFlash 与 TiDB/TiKV 之间 CastStringAsDecimal 行为不一致的问题。
    - 修复由于 "different types: expected Nullable(Int64), got Int64" 错误而导致查询报错的问题
    - 修复由于 "Unexpected type of column: Nullable(Nothing)" 错误而导致的查询报错
    - 修复 Decimal 类型比较时，可能导致数据溢出并导致查询失败的问题
```

Reviewer 7: LGTM

        (dup) - 修复当 TiKV 向同一 Region 发送重复请求时，TiCDC 进程 Panic 的问题 [#2386](https://github.com/pingcap/tiflow/issues/2386)
        (dup) - 修复在多个 TiKV 崩溃或强制重启时可能遇到复制中断的问题 [#3288](https://github.com/pingcap/tiflow/issues/3288)
        (dup) - 修复监控 checkpoint lag 出现负值的问题 [#3010](https://github.com/pingcap/tiflow/issues/3010)
        - 修复 MySQL sink 模块出现死锁时告警过于频繁的问题 [#2706](https://github.com/pingcap/tiflow/issues/2706)
        - 修复 Avro sink 模块不支持解析 JSON 类型列的问题 [#3624](https://github.com/pingcap/tiflow/issues/3624)
        - 修复 TiKV owner 重启时 TiCDC 从 TiKV 读取到错误的元数据快照的问题 [#2603](https://github.com/pingcap/tiflow/issues/2603)
        (dup) - 修复执行 DDL 后的内存泄漏的问题 [#3174](https://github.com/pingcap/tiflow/issues/3174)
        - 修复 Canal 和 Maxwell 协议下 TiCDC 没有自动开启 `enable-old-value` 选项的问题 [#3676](https://github.com/pingcap/tiflow/issues/3676)
        - 修复 `cdc server` 命令在 Red Hat Enterprise Linux 系统的部分版本（如 6.8、6.9 等）上运行时出现时区错误的问题  [#3584](https://github.com/pingcap/tiflow/issues/3584)
        - 修复当 Kafka 为下游时 `txn_batch_size` 监控指标数据不准确的问题 [#3431](https://github.com/pingcap/tiflow/issues/3431)

>          (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)

Reviewer 4: ```suggestion
```

         (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)
        (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)
        (dup) - 修复 tikv_cdc_min_resolved_ts_no_change_for_1m 监控在没有 changefeed 的情况下持续更新的问题 [#11017](https://github.com/tikv/tikv/issues/11017)
        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)

    + (Backup & Restore) BR

        (dup) - 增强恢复的鲁棒性 [#27421](https://github.com/pingcap/tidb/issues/27421)
        - 修复使用了 expression index 的表使用本地 backend 导入时出现错误问题. [#1404](https://github.com/pingcap/br/issues/1404)

>         - 修复 BR 中平均速度不准确的问题 [#1405](https://github.com/pingcap/br/issues/1405)

Reviewer 6: ```suggestion
        - 修复 BR 中平均速度 （average-speed）不准确的问题 [#1405](https://github.com/pingcap/br/issues/1405)
```

    + Dumpling

>         - 修复 dumping 包含 primary/unique key 表时出现的过慢问题. [#29386](https://github.com/pingcap/tidb/issues/29386)

Reviewer 8: ```suggestion
        - 修复 Dumpling 在导出包含复合主键或唯一约束的表时导出速度过慢的问题 [#29386](https://github.com/pingcap/tidb/issues/29386)
```

>         - 修复 dumping 包含 primary/unique key 表时出现的过慢问题. [#29386](https://github.com/pingcap/tidb/issues/29386)

Reviewer 6: ```suggestion
        - 修复 Dumping 包含 primary/unique key 表时，导出速度过慢的问题 [#29386](https://github.com/pingcap/tidb/issues/29386)
```

>         - 修复 TiDB Lightning 在 Local 后端模式下导入使用了表达式索引的表时出现错误的问题 [#1404](https://github.com/pingcap/br/issues/1404)
>         - 修复 BR 中平均速度 (average-speed) 不准确的问题 [#1405](https://github.com/pingcap/br/issues/1405)

Author: ```suggestion
        - 修复 BR 中平均速度 (average-speed) 不准确的问题 [#1405](https://github.com/pingcap/br/issues/1405)
```

---
title: TiDB 5.0.6 Release Notes
---

# TiDB 5.0.6 Release Notes

发版日期：2021 年 12 月 28 日

TiDB 版本：5.0.6

## 兼容性更改

+ Tools

    + TiCDC

        - cdc server 命令错误输出从标准输出改成标准错误. [#3133](https://github.com/pingcap/tiflow/issues/3133)

## 提升改进

+ TiDB

    (dup) - 修复乐观事务冲突可能导致事务相互阻塞的问题 [#11148](https://github.com/tikv/tikv/issues/11148)
    (dup) - 修复运行 MPP 查询时出现 `invalid cop task execution summaries length` 相关日志的问题 [#1791](https://github.com/pingcap/tics/issues/1791)
    (dup) - 当 coprocessor 遇到锁时，在调试日志中显示受影响的 SQL 语句帮助诊断问题 [#27718](https://github.com/pingcap/tidb/issues/27718)

+ TiKV

    - 将插入 SST 文件时的校验操作从 Apply 线程池移动到 Import 线程池，从而提高 SST 文件的插入速度 [#11239](https://github.com/tikv/tikv/issues/11239)

>     - 在 Raft 日志垃圾回收模块中添加了更多监控，从而定位该模块中出现的性能问题 [#11374](https://github.com/tikv/tikv/issues/11374)

Reviewer 4: ```suggestion
    - 在 Raft 日志垃圾回收模块中添加了更多监控指标，从而定位该模块中出现的性能问题 [#11374](https://github.com/tikv/tikv/issues/11374)
```

    - 折叠了部分 Storage 相关的不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)

+ PD

    (dup) - 优化调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)

>     - 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4116](https://github.com/tikv/pd/pull/4116)

Author: ```suggestion
    - 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4497](https://github.com/tikv/pd/issues/4497)
```

>     - 优化调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)

Reviewer 2: ```suggestion
    - 优化 Leader 选举后调度器退出的速度 [#4146](https://github.com/tikv/pd/issues/4146)
```

>     - 通过允许空 Region 调度和修复配置的方式，使 `scatter-range-scheduler` 调度器运行得更加均匀 [#4497](https://github.com/tikv/pd/issues/4497)

Reviewer: ```suggestion
    - 通过允许 `scatter-range-scheduler` 调度器调度空 Region 调度并修复配置，使其调度器运行得更加均匀 [#4497](https://github.com/tikv/pd/issues/4497)
```

+ Tools

    + TiCDC

        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)
        (dup) - 修复当扫描存量数据耗时过长时，可能由于 TiKV 进行 GC 而导致存量数据扫描失败的问题 [#2470](https://github.com/pingcap/tiflow/issues/2470)
        (dup) - 修复当发生 ErrGCTTLExceeded 错误时，changefeed 不快速失败的问题 [#3111](https://github.com/pingcap/tiflow/issues/3111)
        (dup) - 为 EtcdWorker 添加 tick 频率限制，防止 PD 的 etcd 写入次数过于频繁影响 PD 服务 [#3112](https://github.com/pingcap/tiflow/issues/3112)
        - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)
        (dup) - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)

>         - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3352](https://github.com/pingcap/tiflow/issues/3352)

Reviewer 8: ```suggestion
        - Kafka sink 模块添加默认的元数据获取超时时间配置 (`config.Metadata.Timeout`) [#3352](https://github.com/pingcap/tiflow/issues/3352)
```

>         - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3081](https://github.com/pingcap/tiflow/issues/3081)

Author: ```suggestion
```

        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#4054](https://github.com/pingcap/tiflow/issues/4054), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606)

>     + (Backup & Restore) BR

Reviewer 8: ```suggestion
    + Backup & Restore (BR)
```

>         - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#3834](https://github.com/pingcap/tiflow/pull/3834), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606), "go gc", "go_max_procs" [#2998](https://github.com/pingcap/tiflow/pull/2998), "cached region" [#2733](https://github.com/pingcap/tiflow/pull/2733).

Author: ```suggestion
        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#4054](https://github.com/pingcap/tiflow/issues/4054), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606)
```

    - 修复在某些场景下执行 `ALTER TABLE.. ADD INDEX` 语句时 TiDB panic 的问题 [#27687](https://github.com/pingcap/tidb/issues/27687)
    - 修复配置项 `enforce-mpp` 在 v5.0.4 中不生效的问题 [#29252](https://github.com/pingcap/tidb/issues/29252)
    (dup) - 修复当 `CASE WHEN` 函数和 `ENUM` 类型一起使用时的崩溃问题 [#29357](https://github.com/pingcap/tidb/issues/29357)
    (dup) - 修复 `microsecond` 函数的向量化表达式版本结果不正确的问题 [#29244](https://github.com/pingcap/tidb/issues/29244)
    (dup) - 修复 `auto analyze` 输出的日志信息不完整的问题 [#29188](https://github.com/pingcap/tidb/issues/29188)
    (dup) - 修复 `hour` 函数在向量化表达式中执行结果错误的问题 [#28643](https://github.com/pingcap/tidb/issues/28643)
    (dup) - 修复当不支持的 `cast` 被下推到 TiFlash 时出现的非预期错误，例如 `tidb_cast to Int32 is not supported` [#23907](https://github.com/pingcap/tidb/issues/23907)
    (dup) - 修复 MPP 节点的可用性检测在某些边界场景中无法工作的问题 [#3118](https://github.com/pingcap/tics/issues/3118)
    (dup) - 修复分配 `MPP task ID` 时出现 `DATA RACE` 的问题 [#27952](https://github.com/pingcap/tidb/issues/27952)
    (dup) - 修复删除空的 `dual table` 后 MPP 查询出现 `index out of range` 报错的问题 [#28250](https://github.com/pingcap/tidb/issues/28250)

>     - 修复并行插入无效时间类型值时，TiDB panic 的问题 [#25393](https://github.com/pingcap/tidb/issues/25393)

Reviewer 9: ```suggestion
    - 修复并行插入无效日期类型值时，TiDB panic 的问题 [#25393](https://github.com/pingcap/tidb/issues/25393)
```

>     - 修复在 MPP 模式下查询时，报`can not found column in Schema column` 错误的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)

Reviewer 9: ```suggestion
    - 修复在 MPP 模式下查询时，报 `can not found column in Schema column` 错误的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)
```

>     - 修复在 MPP 模式下查询时，报 `can not found column in Schema column` 错误的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)

Author: ```suggestion
    - 修复在 MPP 模式下查询时，报 `can not found column in Schema column` 错误的问题 [#30980](https://github.com/pingcap/tidb/issues/30980)
```

    (dup) - 修复将 `Decimal` 转为 `String` 时长度信息错误的问题 [#29417](https://github.com/pingcap/tidb/issues/29417)
    - 修复 SQL 语句中带有 NATURAL JOIN 时，执行意外报错 `Column 'col_name' in field list is ambiguous` 的问题 [#25041](https://github.com/pingcap/tidb/issues/25041)
    (dup) - 修复由于 `tidb_enable_vectorized_expression` 设置的值不同（`on` 或 `off`）导致 `GREATEST` 函数返回结果不一致的问题 [#29434](https://github.com/pingcap/tidb/issues/29434)
    (dup) - 修复在某些情况下 Planner 可能缓存无效 `join` 计划的问题 [#28087](https://github.com/pingcap/tidb/issues/28087)
    - 修复在某些情况下 SQL 语句在 join 结果上计算聚合函数时，报错 `index out of range [1] with length 1` 的问题 [#1978](https://github.com/pingcap/tics/issues/1978)

+ TiKV

    - 修复某个 TiKV 节点停机会导致 Resolved Timestamp 进度落后的问题 [#11351](https://github.com/tikv/tikv/issues/11351)
    (dup) - 修复 Raft client 中 batch 消息过大的问题 [#9714](https://github.com/tikv/tikv/issues/9714)
    (dup) - 修复在极端情况下同时进行 Region Merge、ConfChange 和 Snapshot 时，TiKV 会出现 Panic 的问题 [#11475](https://github.com/tikv/tikv/issues/11475)
    - 修复逆序扫表时 TiKV 无法正确读到内存锁的问题 [#11440](https://github.com/tikv/tikv/issues/11440)
    (dup) - 修复 Decimal 除法计算的结果为 0 时符号为负的问题 [#29586](https://github.com/pingcap/tidb/issues/29586)

>     - 修复垃圾回收任务的堆积可能会导致 TiKV 内存耗尽的问题 [#11410](https://github.com/tikv/tikv/issues/11410)

Reviewer: @Little-Wallace 可以请帮忙看下这个 issue 对应的内容是否为这条 notes 吗？

    (dup) - 修复 TiKV 监控项中实例级别 gRPC 的平均延迟时间不准确的问题 [#11299](https://github.com/tikv/tikv/issues/11299)
    (dup) - 修复因统计线程监控数据导致的内存泄漏 [#11195](https://github.com/tikv/tikv/issues/11195)
    (dup) - 修复在缺失下游数据库时出现 TiCDC Panic 的问题 [#11123](https://github.com/tikv/tikv/issues/11123)

>     (dup) - 修复因 Congest 错误而导致的 CDC 频繁增加 scan 重试的问题 [#11082](https://github.com/tikv/tikv/issues/11082)

Reviewer 4: ```suggestion
    (dup) - 修复因 Congest 错误而导致的 TiCDC 频繁增加 scan 重试的问题 [#11082](https://github.com/tikv/tikv/issues/11082)
```

    (dup) - 修复因 channel 打满而导致的 Raft 断连的问题 [#11047](https://github.com/tikv/tikv/issues/11047)

>     - 修复在 Lightning 导入数据时 TiKV 会因为文件不存在而出现 Panic 的问题 [#10438](https://github.com/tikv/tikv/issues/10438)

Reviewer 4: ```suggestion
    - 修复在 TiDB Lightning 导入数据时 TiKV 会因为文件不存在而出现 Panic 的问题 [#10438](https://github.com/tikv/tikv/issues/10438)
```

     (dup) - 修复由于无法在 `Max`/`Min` 函数中正确识别 Int64 是否为有符号整数，导致  `Max`/`Min` 函数的计算结果不正确的问题 [#10158](https://github.com/tikv/tikv/issues/10158)
    - 修复从副本从主副本获取快照后未能正确修改元信息导致 TiKV 节点宕机的问题。  [#10225](https://github.com/tikv/tikv/issues/10225)
    - 修复 backup 线程池泄漏的问题 [#10287](https://github.com/tikv/tikv/issues/10287)
    - 修复将不合法字符串转化为浮点数的问题 [#23322](https://github.com/pingcap/tidb/issues/23322)

+ PD

    (dup) - 修复 TiKV 节点缩容后可能导致 Panic 的问题 [#4344](https://github.com/tikv/pd/issues/4344)

>     (dup) - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)

Reviewer: ```suggestion
    - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)
```

Reviewer: @lhy1024 可以请帮忙看下是否可以这么表达吗？

Reviewer 2: 或者 停止服务的节点？

    (dup) - 修复因 Region syncer 卡住而导致 leader 选举慢的问题 [#3936](https://github.com/tikv/pd/issues/3936)
    (dup) - 允许 Evict Leader 调度器调度拥有不健康副本的 Region [#4093](https://github.com/tikv/pd/issues/4093)
    - 修复当对宕机的节点进行修复时删除副本的速度会受限的问题 [#4090](https://github.com/tikv/pd/issues/4090)
    - 修复当 Region 心跳低于 60 秒时热点 Cache 不能清空的问题 [#4390](https://github.com/tikv/pd/issues/4390)

+ TiFlash

>     - 修复扩大整数类型主键的列类型之后，可能会导致数据不一致的问题

Author: ```suggestion
    - 修复整数类型主键的列类型设置为较大范围后数据可能不一致的问题
```

>     - 修复在某些平台 TiFlash 由于找不到 "libnsl.so" 而无法启动的问题，比如某些 ARM 平台

Reviewer 9: ```suggestion
    - 修复在某些平台 TiFlash 由于找不到 `libnsl.so` 库而无法启动的问题，比如某些 ARM 平台
```

    - 修复 TiFlash 的 Store size 统计信息与实际容量不一致的问题

>     - 修复 TiFlash 由于 "Cannot open file" 错误而导致的进程失败

Reviewer 9: ```suggestion
    - 修复 TiFlash 由于 `Cannot open file` 错误而导致的进程失败
```

>     - 修复当查询被取消时，TiFlash 偶发的崩溃问题

Reviewer 9: ```suggestion
    - 修复当 MPP 查询被终止时，TiFlash 偶发的崩溃问题
```

Reviewer 9: 英文是 an MPP query is killed

    - 修复查询报错 "3rd arguments of function substringUTF8 must be constants"

>     - 修复由于过多 "OR" 条件导致的查询报错

Reviewer 9: ```suggestion
    - 修复由于过多 `OR` 条件导致的查询报错
```

>     - 修复 where <string> 查询结果出错的问题

Reviewer 9: ```suggestion
    - 修复 `where <string>` 查询结果出错的问题
```

>     - 修复在 TiFlash 与 TiDB/TiKV 之间，关于 CastStringAsDecimal 行为不一致的问题。

Reviewer 7: ```suggestion
    - 修复在 TiFlash 与 TiDB/TiKV 之间，由于 CastStringAsDecimal 行为不一致的问题。
```

>     - 修复在 TiFlash 与 TiDB/TiKV 之间 CastStringAsDecimal 行为不一致的问题。

Reviewer 9: ```suggestion
    - 修复在 TiFlash 与 TiDB/TiKV 之间 `CastStringAsDecimal` 行为不一致的问题。
```

    - 修复由于 "different types: expected Nullable(Int64), got Int64" 错误而导致查询报错的问题
    - 修复由于 "Unexpected type of column: Nullable(Nothing)" 错误而导致的查询报错

>     - 修复 Decimal 类型比较时，可能导致数据溢出并导致查询失败的问题

Reviewer 9: ```suggestion
    - 修复 `Decimal` 类型比较时，可能导致数据溢出并导致查询失败的问题
```

+ Tools

    + TiCDC

        (dup) - 修复当开启 `force-replicate` 时，可能某些没有有效索引的分区表被忽略的问题 [#2834](https://github.com/pingcap/tiflow/issues/2834)
        - 修复 cdc cli 接收到非预期参数时截断用户参数，导致用户输入数据丢失问题 [#2303](https://github.com/pingcap/tiflow/issues/2303)

>         - 修复 cdc 调度逻辑过早调用表问题 [#2625](https://github.com/pingcap/tiflow/issues/2625)

Reviewer 6: ```suggestion
```
和刘子雄沟通后，建议删除这条。

        (dup) - 修复监控 checkpoint lag 出现负值的问题 [#3010](https://github.com/pingcap/tiflow/issues/3010)
        - 修复 MySQL sink 模块出现死锁时告警过于频繁问题 [#2706](https://github.com/pingcap/tiflow/issues/2706)
        - 修复 Avro sink 模块不支持 json type 列类型解析 [#3624](https://github.com/pingcap/tiflow/issues/3624)
        - 修复 TiKV 重启时 cdc 读取到错误的元数据快照问题 [#2603](https://github.com/pingcap/tiflow/issues/2603)
        (dup) - 修复执行 DDL 后的内存泄漏的问题 [#3174](https://github.com/pingcap/tiflow/issues/3174)
        - 修复 Canal 和 Maxwell 协议没有自动开启 old value 选项问题 [#3676](https://github.com/pingcap/tiflow/issues/3676)
        - 修复 cdc server 在一些 RHEL 系统(6.8, 6.9 etc)上运行出现的时区错误问题.  [#3584](https://github.com/pingcap/tiflow/issues/3584)
        - 修复 Kafka sink 模块监控变量 txn_batch_size 不准确问题 [#3431](https://github.com/pingcap/tiflow/issues/3431)
         (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)
        (dup) - 将 Kafka Sink `partition-num` 的默认值改为 3，使 TiCDC 更加平均地分发消息到各个 Kafka partition [#3337](https://github.com/pingcap/tiflow/issues/3337)

>         - 修复 EtcdWorker row 监控参数错误问题. [#4000](https://github.com/pingcap/tiflow/pull/4000)

Author: ```suggestion
  ```

        (dup) - 修复 tikv_cdc_min_resolved_ts_no_change_for_1m 监控在没有 changefeed 的情况下持续更新的问题 [#11017](https://github.com/tikv/tikv/issues/11017)
        (dup) - 优化 TiKV 重新加载时的速率限制控制，缓解 changefeed 初始化时 gPRC 的拥堵问题 [#3110](https://github.com/pingcap/tiflow/issues/3110)

    + (Backup & Restore) BR

        (dup) - 增强恢复的鲁棒性 [#27421](https://github.com/pingcap/tidb/issues/27421)
        - 修复使用了 expression index 的表使用本地 backend 导入时出现错误问题. [#1404](https://github.com/pingcap/br/issues/1404)
        - 修复 BR 中平均速度不准确的问题 [#1405](https://github.com/pingcap/br/issues/1405)

    + Dumpling

>         - 修复 dumping 包含 primary/unique key 表时出现的过慢问题. [#29386](https://github.com/pingcap/tidb/issues/29386)

Reviewer 8: ```suggestion
        - 修复在导出包含复合主键或唯一约束的表时出现的过慢问题 [#29386](https://github.com/pingcap/tidb/issues/29386)
```

    (dup) - 修复 TiDB 在 TiFlash 关闭时可能出现 panic 的问题 [#28096](https://github.com/pingcap/tidb/issues/28096)
    - 修复优化器在进行 join reorder 优化时，报 `index out of range` 错误的问题 [#24095](https://github.com/pingcap/tidb/issues/24095)
    (dup) - 修复当 `ENUM` 类型作为 `IF` 或 `CASE WHEN` 等控制函数的参数时，返回结果不正确的问题 [#23114](https://github.com/pingcap/tidb/issues/23114)
    - 修复 `CONCAT(IFNULL(TIME(3))` 计算结果出错的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
    - 修复当 `GREATEST` 和 `LEAST`  函数传入无符号整型值时，计算结果出错的问题 [#30101](https://github.com/pingcap/tidb/issues/30101)
    - 修复当 SQL 语句中存在 JSON 类型列与 `CHAR` 类型列连接时，SQL 出错的问题 [#29401](https://github.com/pingcap/tidb/issues/29401)
    - 修复错误使用 lazy existence 检查和 untouched key optimization 可能导致数据索引不一致的问题 [#30410](https://github.com/pingcap/tidb/issues/30410)
    - 修复窗口函数在使用事务时，计算结果与不使用事务时的计算结果不一致的问题 [#29947](https://github.com/pingcap/tidb/issues/29947)
    - 修复 SQL 语句中包含 `cast(integer as char) union string` 时计算结果出错的问题 [#29513](https://github.com/pingcap/tidb/issues/29513)
    (dup) - 修复将 `Decimal` 转为 `String` 时长度信息错误的问题 [#29417](https://github.com/pingcap/tidb/issues/29417)

>     - 修复 SQL 语句中带有 NATURAL JOIN 时，执行意外报错 `Column 'col_name' in field list is ambiguous` 的问题 [#25041](https://github.com/pingcap/tidb/issues/25041)

Reviewer: ```suggestion
    - 修复 SQL 语句中带有 NATURAL JOIN 时，意外报错 `Column 'col_name' in field list is ambiguous` 的问题 [#25041](https://github.com/pingcap/tidb/issues/25041)
```

    - 避免 MPP 查询时候的日志 `invalid cop task execution summaries length`，以防用户疑惑 [#1791](https://github.com/pingcap/tics/issues/1791)
    - 优化 Coprocessor 的 debug 日志，在遇到锁时，打印语句内容，提高问题定位的效率 [#27718](https://github.com/pingcap/tidb/issues/27718)

+ TiKV

    - 将插入 SST 文件时的校验操作移动到了 import 线程池以提高 SST 文件插入速度. [#11239](https://github.com/tikv/tikv/issues/11239)
    - 为 raft 日志回收模块增加了更多监控以定位问题. [#11374](https://github.com/tikv/tikv/issues/11374)
    - 折叠了部分 Storage 相关的不常用监控 [#11681](https://github.com/tikv/tikv/issues/11681)

+ PD

    - 在 PD 切换 Leader 后加速调度器退出 [#4146](https://github.com/tikv/pd/issues/4146)
    - 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4116](https://github.com/tikv/pd/pull/4116)

+ Tools

    + TiCDC

        - 优化 TiKV 重启时限速控制逻辑 [#3110](https://github.com/pingcap/tiflow/issues/3110)
        - 减少 Sink 模块锁冲突 [#2760](https://github.com/pingcap/tiflow/pull/2760)
        - 向 TiKV 注册的 GC safepoint TTL 延迟至1小时，有效支持增量扫时间过长的任务 [#2470](https://github.com/pingcap/tiflow/issues/2470)
        - 出现 ErrGCTTLExceeded 错误时支持快速失败 [#3111](https://github.com/pingcap/tiflow/issues/3111)
        - 增加限速逻辑控制 EtcdWorker tick 频率 [#3112](https://github.com/pingcap/tiflow/issues/3112)

>         - 支持消息 Batch 操作减少 EtcdWorker tick [#3391](https://github.com/pingcap/tiflow/pull/3391)

Author: ```suggestion
        - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)
```

>         - 为 EtcdWorker 添加 tick 频率限制，防止 PD 的 etcd 写入次数过于频繁影响 PD 服务 [#3112](https://github.com/pingcap/tiflow/issues/3112)
>         - 支持消息 Batch 操作减少 EtcdWorker tick [3112](https://github.com/pingcap/tiflow/issues/3112)

Reviewer 8: ```suggestion
        - 为 EtcdWorker 添加 tick 频率限制，防止 PD 的 etcd 写入次数过于频繁影响 PD 服务 [#3112](https://github.com/pingcap/tiflow/issues/3112)
```

Author: @ran-huang 上次修改我删掉了第 47 条，因为它跟上一行是重复的，不知道为啥修改没生效，我再改一次。

        - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
        - Kafka sink 模块添加默认的元数据获取超时时间配置 (`config.Metadata.Timeout`) [#3352](https://github.com/pingcap/tiflow/issues/3352)

>         - 将参数 `max-message-bytes` 的默认值设置为 `10M`，从而减少可能发生的 Kafka 不能发送消息的问题 [#3081](https://github.com/pingcap/tiflow/issues/3081)

Reviewer 8: 这条和 L18 是重复的吗？ @shichun-0415

Author: > 这条和 L18 是重复的吗？ @shichun-0415

是的，这个改动修改了默认值，所以要在兼容性修改体现，另外它解决了一个问题，所以也是一个功能提升。

>         - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#3834](https://github.com/pingcap/tiflow/pull/3834), "mounter row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "table sink total row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "buffer sink total row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "go gc" [#2998](https://github.com/pingcap/tiflow/pull/2998), "go_max_procs" [#2998](https://github.com/pingcap/tiflow/pull/2998), "cached region" [#2733](https://github.com/pingcap/tiflow/pull/2733).

Reviewer 3: ```suggestion
        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#3834](https://github.com/pingcap/tiflow/pull/3834), "mounter row", "table sink total row", "buffer sink total row" [#1606](https://github.com/pingcap/tiflow/issue/1606), "go gc", "go_max_procs" [#2998](https://github.com/pingcap/tiflow/pull/2998), "cached region" [#2733](https://github.com/pingcap/tiflow/pull/2733).
```

    + (Backup & Restore) BR

        - 支持 pd request 和 TiKV IO 超时错误重试 [#27787](https://github.com/pingcap/tidb/issues/27787)

## Bug 修复

+ TiDB

    - 修复在 grant/revoke 语句中，使用全局标识符 `*.*` 报错的问题 [#29675](https://github.com/pingcap/tidb/issues/29675)

>     - 修复 ADD INDEX 在某些场景下 panic 的问题 [#27687](https://github.com/pingcap/tidb/issues/27687)
>     - 修复配置项 `enforce-mpp` 在 v5.0.4 中不生效的问题 [#29252](https://github.com/pingcap/tidb/issues/29252)

Reviewer 9: ```suggestion
    - 修复在某些场景下执行 `ALTER TABLE.. ADD INDEX` 语句时 TiDB panic 的问题 [#27687](https://github.com/pingcap/tidb/issues/27687)
    - 修复配置项 `enforce-mpp` 在 v5.0.4 中不生效的问题 [#29252](https://github.com/pingcap/tidb/issues/29252)
```

    - 修复 `case-when` 表达式在入参为 `enum` 类型时 panic 的问题 [#29357](https://github.com/pingcap/tidb/issues/29357)
    - 修复 `microsecond` 表达式在向量化模式下，计算结果出错的问题 [#29244](https://github.com/pingcap/tidb/issues/29244)
    - 修复 auto analyze 失败时，日志中未记录 SQL 内容的问题 [#29188](https://github.com/pingcap/tidb/issues/29188)
    - 修复 `hour` 表达式在向量化模式下，计算结果出错的问题  [#28643](https://github.com/pingcap/tidb/issues/28643)
    - 修复查询在 MPP 模式下，报错 `tidb_cast to Int32 is not supported` 的问题 [#23907](https://github.com/pingcap/tidb/issues/23907)
    - 修复在一些临界情况下, MPP 节点可用性检测失效的问题 [#3118](https://github.com/pingcap/tics/issues/3118)
    - 修复分配 MPP 任务 id 时可能出现的 data-race 问题 [#27952](https://github.com/pingcap/tidb/issues/27952)
    - 修复 SQL 中包含 `Union` 且运行在 MPP 模式时报错 `index out of range [-1]` 的问题 [#28250](https://github.com/pingcap/tidb/issues/28250)

>     - 修复并行插入无效时间类型值时，TiDB panic 的问题 [#25393](https://github.com/pingcap/tidb/issues/25393)
>     - 修复查询在 MPP 模式下，报错 `can not found column in Schema column` 的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)

Reviewer 9: ```suggestion
    - 修复并行插入无效时间类型值时，TiDB panic 的问题 [#25393](https://github.com/pingcap/tidb/issues/25393)
    - 修复在 MPP 模式下查询时，报`can not found column in Schema column` 错误的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)
```

    - 修复在 TiFlash 停机时，TiDB 可能会 panic 的问题 [#28096](https://github.com/pingcap/tidb/issues/28096)

>     - 修复优化器在进行 join reorder 优化时，报错 `index out of range` 的问题 [#24095](https://github.com/pingcap/tidb/issues/24095).

Reviewer 9: ```suggestion
    - 修复优化器在进行 join reorder 优化时，报 `index out of range` 错误的问题 [#24095](https://github.com/pingcap/tidb/issues/24095).
```

    - 修复 `if`,`case-when`,`elt` 等表达式在入参为 enum 类型时，结果出错的问题 [#23114](https://github.com/pingcap/tidb/issues/23114)

>     - 修复 `concat(ifnull(time(3))` 计算结果出错的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
>     - 修复 `greatest/least` 表达式在入参包含无符号整型时，计算结果出错的问题 [#30101](https://github.com/pingcap/tidb/issues/30101)

Reviewer 9: ```suggestion
    - 修复 `concat(ifnull(time(3))` 计算结果出错的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
    - 修复当 `greatest` 和 `least`  函数传入无符号整型值时，计算结果出错的问题 [#30101](https://github.com/pingcap/tidb/issues/30101)
```

    (dup) - 修复因 channel 打满而导致的 Raft 断连的问题 [#11047](https://github.com/tikv/tikv/issues/11047)
    - 修复 Lightning 导入数据期间 TiKV 因为文件不存在而宕机的问题 [#10438](https://github.com/tikv/tikv/issues/10438)
     (dup) - 修复由于无法在 `Max`/`Min` 函数中正确识别 Int64 是否为有符号整数，导致  `Max`/`Min` 函数的计算结果不正确的问题 [#10158](https://github.com/tikv/tikv/issues/10158)
    - 修复从副本从主副本获取快照后未能正确修改元信息导致 TiKV 节点宕机的问题。  [#10225](https://github.com/tikv/tikv/issues/10225)
    - 修复 backup 线程池泄漏的问题 [#10287](https://github.com/tikv/tikv/issues/10287)
    - 修复将不合法字符串转化为浮点数的问题 [#23322](https://github.com/pingcap/tidb/issues/23322)

+ PD

    (dup) - 修复 TiKV 节点缩容后可能导致 Panic 的问题 [#4344](https://github.com/tikv/pd/issues/4344)
    (dup) - 修复 Operator 被 Down Store 阻塞的问题 [#3353](https://github.com/tikv/pd/issues/3353)
    (dup) - 修复因 Region syncer 卡住而导致 leader 选举慢的问题 [#3936](https://github.com/tikv/pd/issues/3936)
    (dup) - 允许 Evict Leader 调度器调度拥有不健康副本的 Region [#4093](https://github.com/tikv/pd/issues/4093)
    - 修复 Down Store 的 Remove Peer 速度受限问题[#4090](https://github.com/tikv/pd/issues/4090)
    - 修复当 Region 心跳低于60秒导致热点 Cache 不能清空的问题 [#4390](https://github.com/tikv/pd/issues/4390)

+ TiFlash

>     - 修复当在同时使用且扩大主键列类型时，潜在的数据不一致问题。

Reviewer 7: ```suggestion
    - 修复当在扩大整数类型主键的列类型之后，可能会导致数据不一致的问题。
```

>     - 修复当 ARM 架构的机器在使用系统的 libnsl.so 编译 TiFlash 时的错误，修改 INSTALL_DIR 编译 libnsl.so 的选项。

Reviewer 7: ```suggestion
    - 修复在某些平台 TiFlash 由于找不到 "libnsl.so" 而无法启动的问题，比如某些 ARM 平台
```

>     - 修复 TiFlash 统计面板中的 store size 不准确的问题，当 PageStorage 中类型为 checkpoint 的文件删除时，没有减少空间。

Reviewer 7: ```suggestion
    - 修复 TiFlash 的 Store size 统计信息与实际容量不一致的问题
```

>     - 修复当 PageStorage GC 时会误删新创建的 PageFile 导致的异常。

Reviewer 7: ```suggestion
    - 修复 TiFlash 由于 "Cannot open file" 错误而导致的进程失败
```

>     - 修复当用户更新 TiFlash 版本后，同步表时发生的异常。

Reviewer 7: ```suggestion
```

    - 修复当查询被取消时，TiFlash 偶发的崩溃问题。

>     - 修复在 substringUTF8 中的常数异常。

Reviewer 7: ```suggestion
    - 修复查询报错 "3rd arguments of function substringUTF8 must be constants"
```

>     - 当 DAG 请求失败后，加了一个重试 DAG 请求。

Reviewer 7: ```suggestion
    - 提高了支持的查询 SQL 中表达式的层级数量，避免比如过多 "OR" 条件导致查询报错
```

    - 修复当查询带了 where <string> 会变成 int 类型的导致查询结构出错的问题。

>     - 修复 main_capacity_quota_ 变量的检查问题。

Reviewer 7: ```suggestion
```

    - 修复在 TiFlash 与 TiDB/TiKV 之间，关于 CastStringAsDecimal 行为不一致的问题。

>     - 修复 Nullable(Int64) 与 Int64 类型不一致的异常，同步表后以避免类型不一致。

Reviewer 7: ```suggestion
    - 修复由于 "different types: expected Nullable(Int64), got Int64" 错误而导致查询报错的问题
```

>     - 修复列类型为 Nullable(Nothing) 导致的异常。

Reviewer 7: ```suggestion
    - 修复由于 "Unexpected type of column: Nullable(Nothing)" 错误而导致的查询报错
```

>     - 修复 Decimal 类型比较时，可能导致数据溢出并比较失败的问题。

Reviewer 7: ```suggestion
    - 修复 Decimal 类型比较时，可能导致数据溢出并导致查询失败的问题
```

+ Tools

    + TiCDC

        - 修复分区表没有唯一索引时无法复制下发表数据问题 [#2834](https://github.com/pingcap/tiflow/issues/2834)
        - 修复 cdc cli 接收到非预期参数时截断用户参数，导致用户输入数据丢失问题 [#2303](https://github.com/pingcap/tiflow/issues/2303)
        - 修复 cdc 调度逻辑过早调用表问题 [#2625](https://github.com/pingcap/tiflow/issues/2625)
        - 修复 Kafka asyncClient 出错导致的 Kafka_producer 死锁问题 [#2978](https://github.com/pingcap/tiflow/issues/2978)
        - 修复 MQ sink 模块不支持非 binary json 类型列解析 [#2758](https://github.com/pingcap/tiflow/issues/2758)
        - 修复 Kafka sink 模块因为 `max-message-size` 参数过大导致的消息无法发送问题 [#2962](https://github.com/pingcap/tiflow/issues/2962)
        - 修复 TiKV 发生 region merge 导致的锁解析逻辑无法触发问题 [#3061](https://github.com/pingcap/tiflow/issues/3061)
        - 修复 region 重复请求问题[#2386](https://github.com/pingcap/tiflow/issues/2386)
        - 修复 TiKV 崩溃或者强制重启时出现的 region 异常丢失问题 [#3288](https://github.com/pingcap/tiflow/issues/3288)
        - 修复 changefeed checkpoint lag 监控指标出现负值的问题 [#3010](https://github.com/pingcap/tiflow/issues/3010)

>         - 修复 MySQL sink 模块出现死锁时告警过于频繁问题 [#2706](https://github.com/pingcap/tiflow/issues/2706)

Reviewer 8: ```suggestion
        - 修复 MySQL sink 模块出现死锁时告警过于频繁的问题 [#2706](https://github.com/pingcap/tiflow/issues/2706)
```

>         - 修复 Avro sink 模块不支持 json type 列类型解析 [#3624](https://github.com/pingcap/tiflow/issues/3624)

Reviewer 8: ```suggestion
        - 修复 Avro sink 模块不支持解析 JSON 类型列的问题 [#3624](https://github.com/pingcap/tiflow/issues/3624)
```

>         - 修复 TiKV 重启时 cdc 读取到错误的元数据快照问题 [#2603](https://github.com/pingcap/tiflow/issues/2603)

Reviewer 8: ```suggestion
        - 修复 TiKV owner 重启时 TiCDC 从 TiKV 读取到错误的元数据快照的问题 [#2603](https://github.com/pingcap/tiflow/issues/2603)
```

---
title: TiDB 5.0.6 Release Notes
---

# TiDB 5.0.6 Release Notes

发版日期：2021 年 12 月 28 日

TiDB 版本：5.0.6

> 

Reviewer 10: ## Bug 修复
- 修复在 grant/revoke 语句中，使用全局标识符 `*.*` 报错的问题 [#30161](https://github.com/pingcap/tidb/pull/30161)
- 修复 ADD INDEX 在某些场景下 panic 的问题 [#30125](https://github.com/pingcap/tidb/pull/30125)
- 修复配置项 `enforce-mpp` 在 v5.0.4 中不生效的问题 [#29637](https://github.com/pingcap/tidb/pull/29637)
- 修复 `case-when` 表达式在入参为 `enum` 类型时 panic 的问题 [#29509](https://github.com/pingcap/tidb/pull/29509)
- 修复 `microsecond` 表达式在向量化模式下，计算结果出错的问题 [#29385](https://github.com/pingcap/tidb/pull/29385)
- 修复 auto analyze 失败时，日志中未记录 SQL 内容的问题 [#29228](https://github.com/pingcap/tidb/pull/29228)
- 修复 `hour` 表达式在向量化模式下，计算结果出错的问题  [#28871](https://github.com/pingcap/tidb/pull/28871)
- 修复查询在 MPP 模式下，报错 `tidb_cast to Int32 is not supported` 的问题 [#28651](https://github.com/pingcap/tidb/pull/28651)
- 修复在一些临界情况下, MPP 节点可用性检测失效的问题 [#28287](https://github.com/pingcap/tidb/pull/28287)
- 修复分配 MPP 任务 id 时可能出现的 data-race 问题 [#28284](https://github.com/pingcap/tidb/pull/28284)
- 修复 SQL 中包含 `Union` 且运行在 MPP 模式时报错 `index out of range [-1]` 的问题 [#28278](https://github.com/pingcap/tidb/pull/28278)
- 修复并行插入无效时间类型值时，TiDB panic 的问题 [#28198](https://github.com/pingcap/tidb/pull/28198)
- 修复查询在 MPP 模式下，报错 `can not found column in Schema column` 的问题 [#28147](https://github.com/pingcap/tidb/pull/28147)
- 修复在 TiFlash 停机时，TiDB 可能会 panic 的问题 [#28138](https://github.com/pingcap/tidb/pull/28138)
- 修复优化器在进行 join reorder 优化时，报错 `index out of range` 的问题 [#30881](https://github.com/pingcap/tidb/pull/30881). 
- 修复 `if`,`case-when`,`elt` 等表达式在入参为 enum 类型时，结果出错的问题 [#30857](https://github.com/pingcap/tidb/pull/30857)
- 修复 `concat(ifnull(time(3))` 计算结果出错的问题 [#30830](https://github.com/pingcap/tidb/pull/30830)
- 修复 `greatest/least` 表达式在入参包含无符号整型时，计算结果出错的问题 [#30791](https://github.com/pingcap/tidb/pull/30791)
- 修复当 SQL 中存在 json 类型列 与 char 类型列 join 时，SQL 出错的问题 [#30778](https://github.com/pingcap/tidb/pull/30778)
- 修复 update 语句未更新索引处理优化，可能导致数据索引不一致的问题 [#30533](https://github.com/pingcap/tidb/pull/30533)
- 修复窗口函数在使用事务时，计算结果与不使用事务不一样的问题 [#30389](https://github.com/pingcap/tidb/pull/30389)
- 修复 SQL 形如`cast(integer as char) union string` 计算结果出错的问题 [#30055](https://github.com/pingcap/tidb/pull/30055)
- 修复 `concat(decimal_col)` 结果出错的问题 [#30014](https://github.com/pingcap/tidb/pull/30014)
- 修复使用 NATURAL JOIN 时，报错 `Column 'col_name' in field list is ambiguous` 的问题 [#30049](https://github.com/pingcap/tidb/pull/30049)
- 修复 `greatest` 表达式在开启及关闭 `tidb_enable_vectorized_expression` 时，计算结果不同的问题 [#29917](https://github.com/pingcap/tidb/pull/29917)
- 修复优化器在某些 join 场景下缓存无效的执行计划的问题 [#28445](https://github.com/pingcap/tidb/pull/28445)
- 修复 SQL 在 join 上计算聚合函数时，报错 `index out of range [1] with length 1` 的问题 [#28295](https://github.com/pingcap/tidb/pull/28295)
## 提升改进
- 优化乐观事务写入冲突死锁情况下的回滚流程，降低阻塞延迟 [#29776](https://github.com/pingcap/tidb/pull/29776)
- 避免 MPP 查询时候的日志 `invalid cop task execution summaries length`，以防用户疑惑 [#28262](https://github.com/pingcap/tidb/pull/28262)
- 优化 Coprocessor 的 debug 日志，在遇到锁时，打印语句内容，提高问题定位的效率 [#27927](https://github.com/pingcap/tidb/pull/27927)

## 兼容性更改

> 

Reviewer 3: ```suggestion

+ Tools

    + TiCDC
    
        - cdc server 命令错误输出从标准输出改成标准错误. [#3875](https://github.com/pingcap/tiflow/pull/3875)
```

## 功能增强

> ## 提升改进

Reviewer 5: ## TiKV

- 为 raft 日志回收模块增加了更多监控以定位问题. [#11381](https://github.com/tikv/tikv/pull/11381)

> 

Reviewer 2: ```suggestion
+ PD
- 在 PD 切换 Leader 后加速调度器退出 [#4201](https://github.com/tikv/pd/pull/4201)
- 通过允许空 Region 调度和修复配置的方式使 scatter-range-scheduler运行更好 [#4116](https://github.com/tikv/pd/pull/4116)
```

> 

Reviewer 3: + Tools

    + TiCDC

        - 优化 TiKV 重启时限速控制逻辑 [#3132](https://github.com/pingcap/tiflow/pull/3132)
        - 减少 Sink 模块锁冲突 [#2760](https://github.com/pingcap/tiflow/pull/2760)
        - 向 TiKV 注册的 GC safepoint TTL 延迟至1小时，有效支持增量扫时间过长的任务 [#2852](https://github.com/pingcap/tiflow/pull/2852)
        - 出现 ErrGCTTLExceeded 错误时支持快速失败 [#3135](https://github.com/pingcap/tiflow/pull/3135)
        - 增加限速逻辑控制 EtcdWorker tick 频率 [#3268](https://github.com/pingcap/tiflow/pull/3268)
        - 支持消息 Batch 操作减少 EtcdWorker tick [#3391](https://github.com/pingcap/tiflow/pull/3391)
        - Unified sorter 组件支持 cgroup 限制，控制资源消耗 [#3441](https://github.com/pingcap/tiflow/pull/3441)
        - Kafka sink 模块支持默认的元数据获取超时时间 config.Metadata.Timeout [#3670](https://github.com/pingcap/tiflow/pull/3670)
        - Kafka sink 模块设置 `MaxMessageBytes` 默认值为 1MB [#3107](https://github.com/pingcap/tiflow/pull/3107)
        - 增加更多 Promethous 和 grafana 监控告警参数，包括 "no owner alert" [#3834](https://github.com/pingcap/tiflow/pull/3834), "mounter row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "table sink total row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "buffer sink total row" [#2830](https://github.com/pingcap/tiflow/pull/2830), "go gc" [#2998](https://github.com/pingcap/tiflow/pull/2998), "go_max_procs" [#2998](https://github.com/pingcap/tiflow/pull/2998), "cached region" [#2733](https://github.com/pingcap/tiflow/pull/2733).
        
    + (Backup & Restore) BR
    
        - 支持 pd request 和 TiKV IO 超时错误重试 [#1436](https://github.com/pingcap/br/pull/1436)

> ## Bug 修复

Reviewer 5: TiKV
- 修复某个 TiKV 节点停机导致 Resolve Timestamp 进度落后的问题 [#11538](https://github.com/tikv/tikv/pull/11538)
- 修复多条日志合并在同一条消息导致消息发送失败的问题。 [#11532](https://github.com/tikv/tikv/pull/11532)
- 修复 `region` 合并以及发送快照、成员变更这三种情况同时发生时导致 TiKV 宕机的问题 [#11510](https://github.com/tikv/tikv/pull/11510)
- 修复逆序范围查询时没有正确读到内存锁的问题. [#11453](https://github.com/tikv/tikv/pull/11453)
- Fix negative sign when decimal divide to zero [#11333](https://github.com/tikv/tikv/pull/11333)
- Avoid possible OOM due to the accumulation of GC tasks [#11419](https://github.com/tikv/tikv/pull/11419)
- Fix incorrect metrics  "gRPC average duration by-instance". [#11327](https://github.com/tikv/tikv/pull/11327)
- Fix bug of reporting metrics about destroyed threads to Prometheus. [#11201](https://github.com/tikv/tikv/pull/11201)
- Fix CDC panic due to missing downstream. [#11136](https://github.com/tikv/tikv/pull/11136)
- Fix frequent CDC incremental scan retry due to `Congest` error. [#11090](https://github.com/tikv/tikv/pull/11090)
- Fix channel full could break the raft connection [#11070](https://github.com/tikv/tikv/pull/11070)
- Hide untouched storage commands' metrics in grafana dashboard [#11001](https://github.com/tikv/tikv/pull/11001)
- Fix panic caused by SST NotFound, which is ingested by  Lightning import. [#10740](https://github.com/tikv/tikv/pull/10740)
- copr: fix Max/Min bug when comparing signed and unsigned int64 [#10617](https://github.com/tikv/tikv/pull/10617)
- Fix follower meta corruption in rare cases with more than 4 replicas [#10500](https://github.com/tikv/tikv/pull/10500)
- Fix backup threads leak [#10361](https://github.com/tikv/tikv/pull/10361)
- copr cast invalid utf8 string to real bug fix [#9871](https://github.com/tikv/tikv/pull/9871)

> ## Bug 修复

Reviewer 11: # TiFlash
- 修复当在同时使用且扩大主键列类型时，潜在的数据不一致问题。[#3572] (https://github.com/pingcap/tics/pull/3572)
- 修复当 ARM 架构的机器在使用系统的 libnsl.so 编译 TiFlash 时的错误，修改 INSTALL_DIR 编译 libnsl.so 的选项。 [#3205] (https://github.com/pingcap/tics/pull/3205)
- 修复 TiFlash 统计面板中的 store size 不准确的问题，当 PageStorage 中类型为 checkpoint 的文件删除时，没有减少空间。 [#3187] (https://github.com/pingcap/tics/pull/3187)
- 修复当 PageStorage GC 时会误删新创建的 PageFile 导致的异常。[#3226] (https://github.com/pingcap/tics/pull/3226)
- 修复当用户更新 TiFlash 版本后，同步表时发生的异常。[#2701] (https://github.com/pingcap/tics/pull/2701)
- 修复当查询被取消时，TiFlash 偶发的崩溃问题。[#3448] (https://github.com/pingcap/tics/pull/3448)
- 修复在 substringUTF8 中的常数异常。[#3266] (https://github.com/pingcap/tics/pull/3266)
- 当 DAG 请求失败后，加了一个重试 DAG 请求。[#3678] (https://github.com/pingcap/tics/pull/3678)
- 修复当查询带了 where <string> 会变成 int 类型的导致查询结构出错的问题。[#3477] (https://github.com/pingcap/tics/pull/3477)
- 修复了 main_capacity_quota_ 变量的检查问题。[#3425] (https://github.com/pingcap/tics/pull/3425)
- 修复了在 TiFlash 与 TiDB/TiKV 之间，关于 CastStringAsDecimal 行为不一致的问题。[#3674] (https://github.com/pingcap/tics/pull/3674)
- 修复了 Nullable(Int64) 与 Int64 类型不一致的异常，同步表后以避免类型不一致。[#3369] (https://github.com/pingcap/tics/pull/3369)
- 修复了列类型为 Nullable(Nothing) 导致的异常。[#3376] (https://github.com/pingcap/tics/pull/3376)
- 修复 Decimal 类型比较时，可能导致数据溢出并比较失败的问题。 [#3097] (https://github.com/pingcap/tics/pull/3097)

> ## Bug 修复

Reviewer 2: ```suggestion
## Bug 修复
+ PD
- Fix panic issue after TiKV node scales in with some corner cases [#4379](https://github.com/tikv/pd/pull/4379)
- Fix the issue that operator can get blocked due to down store [#4367](https://github.com/tikv/pd/pull/4367)
    - Fix the issue that PD may not elect leader as soon as leader step down [#4218](https://github.com/tikv/pd/pull/4218)
        - Fix the issue that `evict-leader-scheduler` cannot schedule the regions with unhealthy peers. [#4130](https://github.com/tikv/pd/pull/4130)
	    - Fix the issue that there is the store limit of remove peer of the down store [#4097](https://github.com/tikv/pd/pull/4097)
	    - Fix the issue that the hot cache cannot be cleared when the heartbeat interval is less than 60 [#4432](https://github.com/tikv/pd/pull/4432)
```

> ## Bug 修复

Reviewer 3: ```suggestion
## Bug 修复
+ Tools

    + TiCDC

        - 修复分区表没有唯一索引时无法复制下发表数据问题 [#2864](https://github.com/pingcap/tiflow/pull/2864)
        - 修复 cdc cli 接收到非预期参数时截断用户参数，导致用户输入数据丢失问题 [#2888](https://github.com/pingcap/tiflow/pull/2888)
        - 修复 cdc 调度逻辑过早调用表问题 [#2633](https://github.com/pingcap/tiflow/pull/2633)
        - 修复 Kafka asyncClient 出错导致的 Kafka_producer 死锁问题 [#3016](https://github.com/pingcap/tiflow/pull/3016)
        - 修复 MQ sink 模块不支持非 binary json 类型列解析 [#2782](https://github.com/pingcap/tiflow/pull/2782)
        - 修复 Kafka sink 模块因为 `max-message-size` 参数过大导致的消息无法发送问题 [#3047](https://github.com/pingcap/tiflow/pull/3047)
        - 修复 TiKV 发生 region merge 导致的锁解析逻辑无法触发问题 [#3100](https://github.com/pingcap/tiflow/pull/3100)
        - 修复 region 重复请求问题[#3091](https://github.com/pingcap/tiflow/pull/3091)
        - 修复 TiKV 崩溃或者强制重启时出现的 region 异常丢失问题 [#3291](https://github.com/pingcap/tiflow/pull/3291)
        - 修复 changefeed checkpoint lag 监控指标出现负值的问题 [#3533](https://github.com/pingcap/tiflow/pull/3533)
        - 修复 MySQL sink 模块出现死锁时告警过于频繁问题 [#3797](https://github.com/pingcap/tiflow/pull/3797)
        - 修复 Avro sink 模块不支持 json type 列类型解析 [#3651](https://github.com/pingcap/tiflow/pull/3651) 
        - 修复 TiKV 重启时 cdc 读取到错误的元数据快照问题 [#2610](https://github.com/pingcap/tiflow/pull/2610) 
        - 修复过多 DDL 导致的 OOM 问题和元数据 GC 逻辑异常问题 [#3275](https://github.com/pingcap/tiflow/pull/3275)
        - 修复 Canal 和 Maxwell 协议没有自动开启 old value 选项问题 [#3780](https://github.com/pingcap/tiflow/pull/3780)
        - 修复 cdc server 在一些 RHEL 系统(6.8, 6.9 etc)上运行出现的时区错误问题.  [#3908](https://github.com/pingcap/tiflow/pull/3908)
        - 修复 Kafka sink 模块监控变量 txn_batch_size 不准确问题 [#3820](https://github.com/pingcap/tiflow/pull/3820)
        - 修复 Kafka sink 模块当用户关闭 auto-create-topic 选项时没有检查分区数问题 [#3566](https://github.com/pingcap/tiflow/pull/3566)
        - 修复 Kafka message 过大导致的 broker 错误问题. [#3566](https://github.com/pingcap/tiflow/pull/3566)
        - 修复 EtcdWorker row 监控参数错误问题. [#4000](https://github.com/pingcap/tiflow/pull/4000)
        - 修复没有任何任务时 tikv_cdc_min_resolved_ts_no_change_for_1m 告警规则持续告警问题 [#3024](https://github.com/pingcap/tiflow/pull/3024)
        - 修复 cdc 和 TiKV 出现的 gPRC 连接阻塞导致增量扫慢的问题 [#3132](https://github.com/pingcap/tiflow/pull/3132)
        
    + (Backup & Restore) BR
      
        - 修复 Grpc 错误没有重试问题 [#1438](https://github.com/pingcap/br/pull/1438)
        - 修复 BR 中平均速度不准确的问题 [#1411](https://github.com/pingcap/br/pull/1411)
        - 修复使用了 expression index 的表使用本地 backend 导入时出现错误问题. [#1419](https://github.com/pingcap/br/pull/1419)
       
        
    + Dumpling

        - 修复 dumping 包含 primary/unique key 表时出现的过慢问题. [#399](https://github.com/pingcap/dumpling/pull/399)
 ```
